src/Avalonia.Controls.DataGrid/DataGridDataConnection.cs

```diff
@@ -379,17 +379,36 @@ public object GetDataItem(int index)
 
         public bool GetPropertyIsReadOnly(string propertyName)
         {
-            if (DataType != null)
+            Type propertyType = DataType;
+            if (propertyType != null)
             {
+                if (propertyType.GetIsReadOnly())
+                    return true;
+
                 if (!String.IsNullOrEmpty(propertyName))
                 {
-                    Type propertyType = DataType;
+                    Type? propertyReflectableType = null;
+                    if (typeof(IReflectableType).IsAssignableFrom(propertyType))
+                    {
+                        IEnumerator en = DataSource.GetEnumerator();
+                        if (en.MoveNext() && en.Current is IReflectableType reflectableType)
+                        {
+                            propertyReflectableType = reflectableType.GetTypeInfo();
+                        }
+                    }
                     PropertyInfo propertyInfo = null;
-                    List<string> propertyNames = TypeHelper.SplitPropertyPath(propertyName);
-                    for (int i = 0; i < propertyNames.Count; i++)
+                    List<string> propertyNameParts = TypeHelper.SplitPropertyPath(propertyName);
+                    for (int i = 0; i < propertyNameParts.Count; i++)
                     {
-                        propertyInfo = propertyType.GetPropertyOrIndexer(propertyNames[i], out _);
-                        if (propertyInfo == null || propertyType.GetIsReadOnly() || propertyInfo.GetIsReadOnly())
+                        var propertyNamePart = propertyNameParts[i];
+
+                        if (propertyReflectableType != null)
```

Comment:
This check will succeed for subsequent properties in the property path that do have a different type.